lib/src/chain/async_tree.rs

```diff
@@ -811,7 +811,10 @@ where
         assert!(match (self.input_finalized_index, new_best_block) {
             (Some(f), Some(b)) => self.non_finalized_blocks.is_ancestor(f, b),
             (Some(_), None) => false,
-            (None, Some(_)) => true,
+            (None, Some(b)) => {
+                assert!(self.non_finalized_blocks.contains(b));
```

Comment:
This newly-added asset is not strictly related to the bug, but it might catch similar problems earlier on.